webrender/src/device/gfx.rs

```diff
@@ -749,212 +691,187 @@ impl<B: hal::Backend> Buffer<B> {
-            .unwrap()
-            .into();
-        let memory = device
-            .allocate_memory(mem_type, requirements.size)
-            .unwrap();
-        let buffer = device
-            .bind_buffer_memory(&memory, 0, unbound_buffer)
-            .unwrap();
-        CopyBuffer {
-            memory,
+        let buffer = Buffer::new(
+            device,
+            memory_types,
+            usage,
+            pitch_alignment,
+            1048576 * 128, // 128 * 1MB
+            data_stride,
+        );
+        BufferPool {
             buffer,
             data_stride,
-            data_width,
-            data_height,
-            row_pitch_in_bytes,
-            state: Cell::new(hal::buffer::Access::empty())
+            non_coherent_atom_size,
+            offset: 0,
+            size: 0,
         }
     }
 
-    pub fn update<T: Copy>(
-        &mut self,
-        device: &B::Device,
-        buffer_offset: u64,
-        image_data_width_in_bytes: u64,
-        image_data: &[T],
-        offset_alignment: usize,
-    ) -> usize {
-        //assert!(self.state.get().contains(hal::buffer::Access::HOST_WRITE));
-        let buffer_data_width_in_bytes = self.data_width * self.data_stride;
-        let mut needed_height = image_data_width_in_bytes / buffer_data_width_in_bytes as u64;
-        let last_row_length = image_data_width_in_bytes % buffer_data_width_in_bytes as u64;
-        let range = if last_row_length != 0 {
-            needed_height += 1;
-            buffer_offset ..
-                buffer_offset + ((needed_height - 1) as usize * self.row_pitch_in_bytes) as u64 + last_row_length
-        } else {
-            buffer_offset .. (buffer_offset + (needed_height as usize * self.row_pitch_in_bytes) as u64)
-        };
-
-        let mut data = device
-            .acquire_mapping_writer::<T>(
-                &self.memory,
-                range,
-            )
-            .unwrap();
-
-        for y in 0 .. needed_height as usize {
-            let lower_bound = y * self.data_width;
-            let upper_bound = cmp::min(lower_bound + self.data_width, image_data.len());
-            let row = &(*image_data)[lower_bound .. upper_bound];
-            let dest_base = y * self.row_pitch();
-            assert!(dest_base + row.len() <= data.len());
-            data[dest_base .. dest_base + row.len()].copy_from_slice(row);
-        }
-        device.release_mapping_writer(data);
-        (needed_height as usize * self.row_pitch() + offset_alignment) & !offset_alignment
+    pub fn add<T: Copy>(&mut self, device: &B::Device, data: &[T]) {
+        let buffer_len = data.len() * self.data_stride;
+        assert!(
+            self.offset * self.data_stride + buffer_len < self.buffer.buffer_size,
+            "offset({:?}) * data_stride({:?}) + buffer_len({:?}) < buffer_size({:?})",
+            self.offset, self.data_stride, buffer_len, self.buffer.buffer_size
+        );
+        self.size = self.buffer.update(
```

Comment:
`self.size` is awkward here. Do we need to store it? Why?

Reply:
Usually, the size would be `buffer_len (data.len() * data_stride)`, but the `non_coherent_atom_size` can change it, and instead of calculating it twice, we just return it from `buffer.update`.